### PR TITLE
layers: Fix detection of dependent timeline signal during binary wait

### PR DESCRIPTION
**Pull request by artem-lunarg**
Closes https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/8900

### CODE REVIEW

**Comment by artem-lunarg** [comment]
Looking into thread-sanitizer lock-inversion error. This PR does not introduce new synchronization but test activated new code path.

**Review by spencer-lunarg — APPROVED** [review]

**Comment by artem-lunarg** [comment]
Added separate commit to fix lock-inversion found by thread sanitizer

File: layers/state_tracker/queue_state.cpp [hunk 1]
```diff
@@ -141,21 +141,40 @@ std::optional<vvl::SemaphoreInfo> vvl::Queue::FindTimelineWaitWithoutResolvingSi
...
-    for (auto it = submissions_.rbegin(); it != submissions_.rend() && processed_waits < timeline_wait_count_; ++it) {
-        const vvl::QueueSubmission &submission = *it;
-        if (submission.seq > until_seq) {
-            continue;
-        }
-        for (const auto &wait_info : submission.wait_semaphores) {
-            if (wait_info.semaphore->type == VK_SEMAPHORE_TYPE_TIMELINE) {
-                if (!wait_info.semaphore->HasResolvingTimelineSignal(wait_info.payload)) {
-                    return wait_info;
+    // Run algorithm in two separate steps to avoid lock-inversion with Semaphore::RetireWait:
```
Comment (artem-lunarg) [inline layers/state_tracker/queue_state.cpp:144]: That's the fix for thread sanitizer. From what I checked it's indeed a potential deadlock. This function was added not long ago, but thread sanitizer did not detect any issues previously. It looks like we did not have a test to activate a necessary code path, and the test from this PR did the trick.

File: layers/state_tracker/semaphore_state.cpp [hunk 2]
```diff
@@ -288,6 +288,12 @@ void vvl::Semaphore::GetLastBinarySignalSource(VkQueue &queue, vvl::Func &acquir
 bool vvl::Semaphore::HasResolvingTimelineSignal(uint64_t wait_payload) const {
     assert(type == VK_SEMAPHORE_TYPE_TIMELINE);
     auto guard = ReadLock();
+
+    // Check if completed payload value (which includes initial value) resolves the wait.
+    if (wait_payload <= completed_.payload) {
```
Comment (artem-lunarg) [inline layers/state_tracker/semaphore_state.cpp:293]: That's the main fix for false positive.

**Pull request merged**